### PR TITLE
Add CHIP_ERROR object methods

### PR DESCRIPTION
**Pull request by kpschoedel**
#### Problem

Having CHIP_ERROR be a class type provides type safety and the option to
trace the source of errors. Since all platforms now use the class type,
transitional code that provided for CHIP_ERROR to be either a class or
integer type is no longer required and can be removed, which simplifies
the class and some notation for using errors.

#### Change overview

- Remove the `CHIP_CONFIG_ERROR_CLASS` configuration option.
- Add instance methods to replace the static `ChipError` functions, which
  were present to allow transitional integer overloads. To avoid
  breaking PRs in flight, the static methods remain present for now,
  marked “DO NOT USE”; they will be removed in a final cleanup PR.

#### Testing

Existing tests should confirm no change to functionality.

### CODE REVIEW

File: src/lib/core/CHIPError.h [hunk 1]
```diff
@@ -91,92 +94,248 @@ class ChipError
...
+        return std::numeric_limits<typename std::make_unsigned<T>::type>::digits <= kValueLength;
+    }
+
+    /**
+     * Test whether if @a value can be losslessly encapsulated in a CHIP_ERROR.
+     */
+    template <typename T>
+    static constexpr bool CanEncapsulate(T value)
+    {
+        return CanEncapsulate<T>() || FitsInField(kValueLength, static_cast<ValueType>(value));
```
Comment (Damian-Nordic) [inline src/lib/core/CHIPError.h:286]: Should we also check if `value` can be cast to `ValueType`?

File: src/lib/core/CHIPError.h [hunk 2]
```diff
@@ -91,92 +94,248 @@ class ChipError
...
+    /**
+     * Construct a CHIP_ERROR encapsulating @a value inside the Range @a range.
+     *
+     * @note
+     *  The result is valid only if CanEncapsulate() is true.
+     */
+    constexpr ChipError(Range range, ValueType value) :
+        mError(MakeInteger(range, (value & MakeMask(0, kValueLength)))) CHIP_INITIALIZE_ERROR_SOURCE(nullptr, 0)
+    {}
+    constexpr ChipError(Range range, ValueType value, const char * file, unsigned int line) :
```
[the remark this answers is not in the document]
Comment (kpschoedel) [reply src/lib/core/CHIPError.h:131]: The intent behind always providing both is that a caller doesn't need its own #if.

[the remark this answers is not in the document]
Comment (Damian-Nordic) [reply src/lib/core/CHIPError.h:131]: I see, thanks.

File: src/lib/core/CHIPError.h (hunk 1 above)
Comment (kpschoedel) [reply src/lib/core/CHIPError.h:286]: I _think_ we get that from `CanEncapsulate<T>()`, unless I'm missing some weird C++ corner, assuming `kValueLength` bits fit in `ValueType`. Perhaps I should add a bunch of static_asserts for the bit field constants.

File: src/lib/core/CHIPError.h [hunk 3]
```diff
@@ -91,92 +94,248 @@ class ChipError
...
 
-    /// Construct a `CHIP_ERROR` encapsulating @a value inside the @a range.
-    static constexpr ErrorType Encapsulate(Range range, ValueType value)
+    /**
+     * Test whether type @a T can always be losslessly encapsulated in a CHIP_ERROR.
+     */
+    template <typename T>
+    static constexpr bool CanEncapsulate()
+    {
+        return std::numeric_limits<typename std::make_unsigned<T>::type>::digits <= kValueLength;
```
[the remark this answers is not in the document]
Comment (kpschoedel) [reply src/lib/core/CHIPError.h]: Done

File: src/lib/core/CHIPError.h (hunk 1 above)
Comment (kpschoedel) [reply src/lib/core/CHIPError.h:286]: Added field assertions.

**Review by Damian-Nordic — APPROVED** [review]

**Review by andy31415 — APPROVED** [review]

**Review by yufengwangca — APPROVED** [review]

**Comment by andy31415** [comment]
@kpschoedel - merge conflicts

**Comment by bzbarsky-apple** [comment]
So switching to class-type errors was not a problem in terms of codesize?  I guess as long as `CHIP_CONFIG_ERROR_SOURCE` is false the codesize is fine?  @kpschoedel

**Comment by kpschoedel** [comment]
> So switching to class-type errors was not a problem in terms of codesize? I guess as long as `CHIP_CONFIG_ERROR_SOURCE` is false the codesize is fine? @kpschoedel

Yes, I checked that before embarking on this conversion. With all the relevant things inline the code for a class-wrapped integer is generally identical to a bare one.

**Review by bzbarsky-apple — APPROVED** [review]

File: examples/window-app/efr32/src/main.cpp [hunk 4]
```diff
@@ -65,7 +65,7 @@ extern "C" void vApplicationIdleHook(void)
 
 void appError(CHIP_ERROR err)
 {
-    // appError(static_cast<int>(chip::ChipError::AsInteger(error)));
+    // appError(static_cast<int>(error.AsInteger()));
```
[the remark this answers is not in the document]
Comment (kpschoedel) [reply examples/window-app/efr32/src/main.cpp:68]: I'll add that to the cleanup PR.

**Review by msandstedt — APPROVED** [review]

**Review by mspang — APPROVED** [review]

**Comment by kpschoedel** [comment]
I'll now start monitoring PRs for the obsolete functions and submit a final cleanup when it's safe.

**Pull request merged**